build.zig

```diff
@@ -954,6 +952,99 @@ fn client_docs(
     maybe_execute(b, allocator, client_docs_build, "client_docs");
 }
 
+/// Detects the system's JVM.
+/// JNI tests requires the Java SDK installed and
+/// JAVA_HOME environment variable set.
+const JniTestStep = struct {
+    step: std.build.Step,
+    tests: *std.build.LibExeObjStep,
+
+    fn add(b: *std.build.Builder, mode: Mode, target: CrossTarget) *std.build.LibExeObjStep {
+        const tests = b.addTest("src/clients/java/src/jni_tests.zig");
+        tests.linkSystemLibrary("jvm");
+        tests.linkLibC();
+        tests.setTarget(target);
+        if (builtin.os.tag == .windows) {
+            // TODO(zig): The function `JNI_CreateJavaVM` tries to detect
+            // the stack size and causes a SEGV that is handled by Zig's panic handler.
+            // https://bugzilla.redhat.com/show_bug.cgi?id=1572811#c7
+            //
+            // The workaround is run the tests in "ReleaseFast" mode.
+            tests.setBuildMode(.ReleaseFast);
+        } else {
+            tests.setBuildMode(mode);
+        }
+
+        var jvm_check_step = b.allocator.create(JniTestStep) catch unreachable;
+        jvm_check_step.* = .{
+            .step = std.build.Step.init(.custom, "jni test", b.allocator, JniTestStep.make),
+            .tests = tests,
+        };
+
+        tests.step.dependOn(&jvm_check_step.step);
+        return tests;
+    }
+
+    fn make(step: *std.build.Step) anyerror!void {
+        const self = @fieldParentPtr(JniTestStep, "step", step);
+        const builder: *std.build.Builder = self.tests.builder;
+
+        const java_home = builder.env_map.get("JAVA_HOME") orelse {
+            std.log.err(
+                "JNI tests require the Java SDK installed and JAVA_HOME environment variable set.",
+                .{},
+            );
+            return error.JavaHomeNotSet;
+        };
+
+        const libjvm_path = builder.pathJoin(&.{
+            java_home,
+            if (builtin.os.tag == .windows) "/lib" else "/lib/server",
+        });
+        self.tests.addLibPath(libjvm_path);
+
+        switch (builtin.os.tag) {
+            .windows => set_windows_dll(builder.allocator, java_home),
+            .macos => try builder.env_map.put("DYLD_LIBRARY_PATH", libjvm_path),
+            .linux => {
+                // On Linux, detects the abi by calling `ldd` to check if
+                // the libjvm.so is linked against libc or musl.
+                // It's reasonable to assume that ldd will be present.
+                const abi = builder.exec(&.{
+                    "sh",
+                    "-c",
+                    builder.fmt(
+                        "ldd {s} | grep -q musl && echo musl || echo libc",
+                        .{builder.pathJoin(&.{ libjvm_path, "libjvm.so" })},
+                    ),
+                }) catch unreachable;
+                self.tests.target.abi = if (std.mem.startsWith(u8, abi, "musl")) .musl else .gnu;
```

Comment:
:thinking: would it be correct to set step's `abi` at execution time, rather than at configuration time? I think it would work fine, given the current impl, but I am not sure it'll be guaranteed to work. 

I guess it doesn't really matter, and the current solution seems simple enough.

Reply:
Not only `tests.target.abi` but `tests.addLibPath` is called during the execution phase of the parent step, before the test step runs, so I guess it is still in its configuration phase.

Unless `zig build` changes this behavior in the future to render all commands at once before calling `step.make(...)` on the parent step.